docs/internal/foreign_code.md

```diff
@@ -62,32 +65,210 @@ We hope that the region abstraction will make programming with sandboxes natural
+
+Verona's unified call syntax provides a mechanism for disambiguating fields in C++ classes that are in both a base class and a subclass.
+For example, consider this C++ code:
+
+```c++
+struct Super
 {
-	[#native, #interface_language="c++"]
-	imports library("libLLVM.so", "llvm.h");
-}
+  void x();
+};
+struct Sub : public Super
+{
+  void x();
+};
+```
+
+In Verona, given an instance of `Sub`, you could call `Super::x()` like this:
+
+```verona
+var s = CXXLib.Sub();
+CXXLib.Super.x(CXXLib.static_cast[CXXLib.Sub](s));
+// The following two are equivalent 
+s.x();
+CXXLib.Sub.x(s);
 ```
 
-This includes two attributes that tell the compiler the format of the library (native code) and how to parse the header (which, in this example, would have to be something provided by the user that included the desired headers).
-Once C++ module support is more widespread, module imports can take the place of header inclusions.
+### Templates
+
+C++ templates are exposed as Verona generics.
+Unlike normal Verona generics, C++ templates can take values as type parameters.
+For the MVP, we should not support C++20 object literals and so support only types and numeric types as template parameters.
+This is sufficient for existing rich C++ APIs and requires approximately the same parsing logic as is already needed for Verona arrays.
+
+Unlike Verona generics, a C++ template may parse but not be possible to instantiate with some of the types that it permits as arguments.
+For example, `std::vector<T>` requires that `T` is copy or move constructable but the only constraint on `T` is that it is a type.
+To avoid propagating this problem into Verona code, we require that all template arguments are concrete types *before* reification.
+A Verona generic can wrap a C++ template by providing a closed set of permitted values for the template instantiation.
+If any of these fail, the generic will not valid Verona code, if all of them succeed then the generic is valid for all possible instantiations.
 
-This instructs the compiler to expose the C++ namespace and class structure the `LLVM` Verona class.
-C++ namespaces and classes will appear as nested classes within the Verona `LLVM` class.
-For example, `LLVM.llvm.SmallVector[LLVM.int, 12]` corresponds to `llvm::SmallVector<int, 12>`.
+### Overloaded operators
 
-Instantiating this class would require the user to specify a sandboxing policy, for example (again, straw-man syntax):
+Verona allows overloading most of the operators that C++ provides and so these can be trivially mapped just like any other methods.
+The main exceptions relate to assignment, though Verona's `apply` sugar provides similar functionality.
+Classes from C++ need to opt out of `create` and `apply` sugar because these reflect patterns that don't exist in C++ codebases and so it's possible that apply sugar could be applied for C++ assignment operators.
+
+### Casts
```

Comment:
This section could come before `Classes`, as it has an example of `static_cast` and I assumed other things would fall through the net, which isn't true.